File: KahveQRKodOlusturucuJava-calismalarim/app/src/main/java/com/example/qrkodolusturucujava/viewmodel/MainVM.java
package com.example.qrkodolusturucujava.viewmodel;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.qrkodolusturucujava.data.numara_kayit.NumaraKayitSp;
import com.example.qrkodolusturucujava.model.CRUDCevap;
import com.example.qrkodolusturucujava.services.Services;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@HiltViewModel
public class MainVM extends ViewModel {

    private static final String TAG = "MainVM";

    private final Services kullaniciServisDao;
    private final NumaraKayitSp numaraKayitSp;

    private final MutableLiveData<Integer> hediyeKahveLiveData = new MutableLiveData<>();
    private final MutableLiveData<Integer> kahveSayisiLiveData = new MutableLiveData<>();
    private final MutableLiveData<String> telefonNumarasiLiveData = new MutableLiveData<>();

    public LiveData<Integer> getHediyeKahveLiveData() {
        return hediyeKahveLiveData;
    }

    public LiveData<Integer> getKahveSayisiLiveData() {
        return kahveSayisiLiveData;
    }

    public LiveData<String> getTelefonNumarasiLiveData() {
        return telefonNumarasiLiveData;
    }

    @Inject
    public MainVM(Services kullaniciServisDao, NumaraKayitSp numaraKayitSp) {
        this.kullaniciServisDao = kullaniciServisDao;
        this.numaraKayitSp = numaraKayitSp;
        startPeriodicKullaniciEkle();
    }

    public void kullaniciEkle(String numara) {
        if (numara.isEmpty()) {
            Log.e(TAG, "Telefon numarası boş! Kullanıcı eklenemez.");
            return;
        }

        kullaniciServisDao.kullaniciEkle(numara, new Callback<CRUDCevap>() {
            @Override
            public void onResponse(Call<CRUDCevap> call, Response<CRUDCevap> response) {
                if (response.isSuccessful() && response.body() != null) {
                    CRUDCevap cevap = response.body();
                    Log.d(TAG, "API Yanıtı: " + cevap.getMessage());
                    Log.d(TAG, "API Success: " + cevap.getSuccess());

                    hediyeKahveLiveData.postValue(cevap.getHediyeKahve());
                    kahveSayisiLiveData.postValue(cevap.getKahveSayisi());
                } else {
                    Log.e(TAG, "Yanıt başarısız veya boş");
                }
            }

            @Override
            public void onFailure(Call<CRUDCevap> call, Throwable t) {
                Log.e(TAG, "API isteği başarısız: " + t.getLocalizedMessage());
            }
        });
    }
    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();

    private void startPeriodicKullaniciEkle() {
        scheduler.scheduleAtFixedRate(() -> {
            String numara = numaraKayitSp.numaraGetir();

            if (numara != null && !numara.isEmpty()) {
                kullaniciEkle(numara);
            } else {
                Log.e(TAG, "Numara kaydedilmemiş!");
            }

        }, 0, 15, TimeUnit.SECONDS);
    }

    public boolean checkTelefonNumarasi() {
        String savedPhoneNumber = numaraKayitSp.numaraGetir();
        if (savedPhoneNumber == null || savedPhoneNumber.isEmpty()) {
            Log.e(TAG, "Telefon numarası kaydedilmemiş!");
            return false;
        }
        telefonNumarasiLiveData.postValue(savedPhoneNumber);
        kullaniciEkle(savedPhoneNumber);
        return true;
    }

    public void numarayiKaydet(String numara) {
        numaraKayitSp.numaraKaydet(numara);
        Log.e(TAG, "Numara kaydedildi: " + numara);
        telefonNumarasiLiveData.postValue(numara);
    }


}

File: KahveQrOlusturucuKotlin-calismalarim/app/src/main/java/com/example/qrkodolusturucu/viewmodel/MainVM.kt
package com.example.qrkodolusturucu.viewmodel

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.qrkodolusturucu.data.numara_kayit.NumaraKayitSp
import com.example.qrkodolusturucu.model.CRUDCevap
import com.example.qrkodolusturucu.services.Services
import com.example.qrkodolusturucu.services.ServicesImpl
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import javax.inject.Inject

private const val TAG = "MainVM"
@HiltViewModel
class MainVM @Inject constructor(
    private val kullaniciServisDao: Services,
    private val numaraKayitSp: NumaraKayitSp
) : ViewModel() {

    private val _hediyeKahveLiveData = MutableLiveData<Int>()
    val hediyeKahveLiveData: LiveData<Int> get() = _hediyeKahveLiveData

    private val _kahveSayisiLiveData = MutableLiveData<Int>()
    val kahveSayisiLiveData: LiveData<Int> get() = _kahveSayisiLiveData

    private val _telefonNumarasiLiveData = MutableLiveData<String?>()
    val telefonNumarasiLiveData: LiveData<String?> get() = _telefonNumarasiLiveData

    init {
        startPeriodicKullaniciEkle()
    }

    fun kullaniciEkle(numara: String) {
        if (numara.isEmpty()) {
            Log.e(TAG, "Telefon numarası boş! Kullanıcı eklenemez.")
            return
        }

        viewModelScope.launch {
            try {
                val response = kullaniciServisDao.kullaniciEkle(numara)

                Log.d(TAG, "API Yanıtı: ${response.message}")
                Log.d(TAG, "API Success: ${response.success}")

                    _hediyeKahveLiveData.postValue(response.hediyeKahve)
                    _kahveSayisiLiveData.postValue(response.kahveSayisi)


            } catch (e: Exception) {
                Log.e(TAG, "API isteği başarısız: ${e.localizedMessage}")
            }
        }

    }



    fun checkTelefonNumarasi():Boolean {
        val savedPhoneNumber = numaraKayitSp.numaraGetir()
        if (savedPhoneNumber.isNullOrEmpty()) {
            Log.e(TAG, "Telefon numarası kaydedilmemiş!")
            return false
        }
        // Numarayı aldıktan sonra API çağrısını yap
        _telefonNumarasiLiveData.postValue(savedPhoneNumber)
        kullaniciEkle(savedPhoneNumber)
        return true
    }


    fun numarayiKaydet(numara: String) {
        numaraKayitSp.numaraKaydet(numara)
        Log.e(TAG, "Numara kaydedildi: $numara")  // Burada numaranın doğru kaydedildiğini logluyoruz.
        _telefonNumarasiLiveData.postValue(numara)
    }


    private fun startPeriodicKullaniciEkle() {
        viewModelScope.launch {
            // Başlangıçta numarayı kontrol et
            var numara = numaraKayitSp.numaraGetir()

            // Numara yoksa sürekli kontrol et
            while (true) {
                if (!numara.isNullOrEmpty()) {
                    // Numara varsa, kullanıcıyı ekle
                    kullaniciEkle(numara)
                } else {
                    Log.e(TAG, "Numara kaydedilmemiş!")
                }

                // 15 saniye bekle ve tekrar kontrol et
                delay(15000) // 15 saniye bekle
                numara = numaraKayitSp.numaraGetir() // Yeni numarayı kontrol et
            }
        }
    }


}
